### PR TITLE
CI/CD setup

### PR DESCRIPTION
**Pull request by frank-wei**
# Description
This is PR is not be ready for merge yet. There are some points we need to improve together with NV.  

This PR is for setting up the CI/CD. Pytorch collaborates with circle CI to provide CI/CD. I am new to circle CI so I spent some efforts to read their doc and come up with our 1st version. Since we need to use GPU machine, we actually can only choose their [machine executor](https://circleci.com/docs/2.0/executor-intro#gpu) with choice of 3 types of card [P4,T4,V100](https://circleci.com/docs/2.0/configuration-reference#gpu-executor-linux)
Their available GPU [images](https://circleci.com/docs/2.0/configuration-reference#available-linux-gpu-images) come with cuda driver and their supporting libs installed. We can not use our own docker image as it is not supported (error msg below). 
```
Build-agent version 1.0.129077-32fb56c8 (2022-06-17T08:48:08+0000)
Creating a dedicated VM with nvcr.io/nvidia/tensorrt:22.01-py3 image
failed to create host: Image nvcr.io/nvidia/tensorrt:22.01-py3 is not supported
```

So based on the vanilla cuda environment, here are the steps that I set up in [config.yaml](https://github.com/pytorch/TensorRT/blob/circleci-editor/245/circleci-project-setup/.circleci/config.yml) (recipe to set up the CI/CD)
1. install cudnn, tensorRT lib 
2. install python required libs and tensorRT python lib 
3. python setup.py install 
4. run the fx2trt tests ( [passed](https://app.circleci.com/pipelines/github/pytorch/TensorRT/52/workflows/bb2b8b89-848d-4ea4-9075-58cc4444787e/jobs/62) )
5. tried to run torch_tensorrt tests (failed and need help from @narendasan. Below is the command I used in ssh mode. The circle CI supports ssh model to debug on their machine)

```
# modules
circleci@default---us-central1-a---005145ed-e543-4a5c-bb30-119e0ba80409:~/project$ bazel test //tests/modules:test_modules --compilation_mode=dbg --test_output=errors
ERROR: Skipping '//tests/modules:test_modules': no such target '//tests/modules:test_modules': target 'test_modules' not declared in package 'tests/modules' defined by /home/circleci/project/tests/modules/BUILD
ERROR: no such target '//tests/modules:test_modules': target 'test_modules' not declared in package 'tests/modules' defined by /home/circleci/project/tests/modules/BUILD
 

# core
circleci@default---us-central1-a---005145ed-e543-4a5c-bb30-119e0ba80409:~/project$ bazel test //tests/core:core_tests  --compilation_mode=dbg --test_output=errors --jobs=1
ERROR: /home/circleci/project/tests/core/BUILD:8:10: no such target '//tests/modules:mobilenet_v2_scripted.jit.pt': target 'mobilenet_v2_scripted.jit.pt' not declared in package 'tests/modules' defined by /home/circleci/project/tests/modules/BUILD and referenced by '//tests/core:jit_models'
ERROR: Analysis of target '//tests/core:test_detecting_input_type' failed; build aborted: 
 
# py
circleci@default---us-central1-a---005145ed-e543-4a5c-bb30-119e0ba80409:~/project$ bazel test //tests/py:test_api  --compilation_mode=dbg --test_output=errors --jobs=1
ERROR: error loading package 'tests/py': Unable to find package for @py_test_deps//:requirements.bzl: The repository '@py_test_deps' could not be resolved: Repository '@py_test_deps' is not defined.
```

Overall, circle CI is good to use and can satisfy our needs for general CI/CD. 
Pros:
1. meets general needs 
2. speed is ok

Cons:
1. 3 types of machines P4,T4,V100
2. GPU images are limited 
3. trigger mode supports new PR trigger and fixed time trigger only. Not flexible for certain folder.

I want to discuss with NV folks in general, if we could deploy circle CI for our CI/CD. What else you think we can possibly improve?

### CODE REVIEW

**Comment by yinghai** [comment]
Seems good to me. It's better than no CI. :)

File: WORKSPACE [hunk 1]
```diff
@@ -104,29 +104,29 @@ http_archive(
...
+)
 
-#new_local_repository(
-#    name = "libtorch_pre_cxx11_abi",
-#    path = "/usr/local/lib/python3.6/dist-packages/torch",
-#    build_file = "third_party/libtorch/BUILD"
-#)
+new_local_repository(
+    name = "libtorch_pre_cxx11_abi",
+    path = "/opt/circleci/.pyenv/versions/3.9.4/lib/python3.9/site-packages/torch",
```
Comment (narendasan) [inline WORKSPACE:115]: Do we want to build both variants of the ABI in CI?

File: .circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            BAZEL_VERSION=5.1.1
+            sudo wget -q https://github.com/bazelbuild/bazel/releases/download/${BAZEL_VERSION}/bazel-${BAZEL_VERSION}-linux-x86_64 -O /usr/bin/bazel
+            sudo chmod a+x /usr/bin/bazel
+
+      - run:
+          name: set up python environment
+          command: |
+            pip3 install nvidia-pyindex
+            pip3 install nvidia-tensorrt==8.2.4.2
+            pip3 install --pre torch==1.13.0.dev20220618  torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cu113
```
Comment (narendasan) [inline .circleci/config.yml:47]: For this, cudnn and TensorRT, is there a good way to parameterize this file so we can just maintain a list of versions somewhere?

**Comment by narendasan** [comment]
Also is there a way to have Circle CI skip certain commits. For example we auto generate the documentation and check it into the repo. It would be good to skip those commits.

**Comment by yinghai** [comment]
@narendasan For documentation changes may add `[skip ci]` in PR title and it should just work? 

https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/

**Comment by frank-wei** [comment]
@narendasan  I addressed the comments before. If you are Ok with it, could we land it first and could you help to add the torchscript tests later? We can iterate based on this version as a beginning for our future test needs.

File: py/setup.py [hunk 3]
```diff
@@ -23,11 +23,14 @@
 JETPACK_VERSION = None
 
 __version__ = '1.2.0a0'
-
+FX2TRT_ONLY = False
 
 def get_git_revision_short_hash() -> str:
     return subprocess.check_output(['git', 'rev-parse', '--short', 'HEAD']).decode('ascii').strip()
 
+if "--fx2trt-only" in sys.argv:
```
[the remark this answers is not in the document]
Comment (narendasan) [reply py/setup.py:31]: I think we should drop references to fx2trt for the most part

[the remark this answers is not in the document]
Comment (yinghai) [reply py/setup.py:31]: Yeah exactly. fx-only is a much better name here.

**Review by yinghai — APPROVED** [review]

**Pull request merged**